File: solutions/17.py
class Solution:
    def letterCombinations(self, digits):
        m = {'2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl', '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz'}

        if not digits:
            return []

        if len(digits) == 1:
            return list(m[digits])

        arrs = self.letterCombinations(digits[1:])
        return [ch + arr for ch in m[digits[0]] for arr in arrs]

File: rb/17.rb
# @param {String} digits
# @return {String[]}
def letter_combinations(digits)
  map = {
    "2" => "abc",
    "3" => "def",
    "4" => "ghi",
    "5" => "jkl",
    "6" => "mno",
    "7" => "pqrs",
    "8" => "tuv",
    "9" => "wxyz"
  }

  return [] if digits == ""
  return map[digits[0]].chars if digits.size == 1

  map[digits[0]].chars.flat_map do |ch|
    letter_combinations(digits[1..-1]).map { |ch| ch + rest }
  end
end
